File: _designs_001/Implement Trie (Prefix Tree)/Solution.java
public class Solution {
    public List<Boolean> testTrie(StreamTokenizer tokenizer, List<Double> metrics) throws IOException {
        Trie trie = new Trie();
        List<Boolean> result = new ArrayList<Boolean>();

        int opsCnt = Serializer.deserializeInt(tokenizer);
        for (int i = 0; i < opsCnt; i++) {
            String word;
            double startTime;
            switch (Serializer.deserializeInt(tokenizer)) {
            case 0:
                word = Serializer.deserializeString(tokenizer);
                startTime = System.nanoTime();
                trie.insert(word);
                metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
                break;
            case 1:
                word = Serializer.deserializeString(tokenizer);
                startTime = System.nanoTime();
                result.add(trie.search(word));
                metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
                break;
            case 2:
                word = Serializer.deserializeString(tokenizer);
                startTime = System.nanoTime();
                result.add(trie.startsWith(word));
                metrics.set(0, metrics.get(0) + (double) (System.nanoTime() - startTime) / 1000000L);
                break;
            }
        }
        return result;
    }
}

File: _designs_001/Implement Trie (Prefix Tree)/Solution.rb
class Solution
  def testTrie(inp, metrics)
    metrics[0] = 0
    trie, r = Trie.new, []
    Serializer.deserializeInt(inp).times do
      case Serializer.deserializeInt(inp)
      when 0
        word = Serializer.deserializeString(inp)
        start_time = Time.now
        trie.insert(word)
        metrics[0] += (Time.now - start_time) * 1000
      when 1
        word = Serializer.deserializeString(inp)
        start_time = Time.now
        r << trie.search(word)
        metrics[0] += (Time.now - start_time) * 1000
      when 2
        word = Serializer.deserializeString(inp)
        start_time = Time.now
        r << trie.startsWith(word)
        metrics[0] += (Time.now - start_time) * 1000
      end
    end
    r
  end
end
